controller/src/controller.rs

```diff
@@ -0,0 +1,149 @@
+use super::{
+    error::{self, Result},
+    statemachine::determine_next_node_spec,
+};
+use models::node::{BottlerocketNode, BottlerocketNodeClient, BottlerocketNodeState};
+
+use kube::runtime::reflector::Store;
+use kube::ResourceExt;
+use snafu::ResultExt;
+use std::collections::BTreeMap;
+use tokio::time::{sleep, Duration};
+use tracing::{event, instrument, Level};
+
+// Defines the length time after which the controller will take actions.
+const ACTION_INTERVAL: Duration = Duration::from_secs(2);
+
+/// The BrupopController orchestrates updates across a cluster of Bottlerocket nodes.
+pub struct BrupopController<T: BottlerocketNodeClient> {
+    node_client: T,
+    brn_reader: Store<BottlerocketNode>,
+}
+
+impl<T: BottlerocketNodeClient> BrupopController<T> {
+    pub fn new(node_client: T, brn_reader: Store<BottlerocketNode>) -> Self {
+        BrupopController {
+            node_client,
+            brn_reader,
+        }
+    }
+
+    /// Returns a list of all `BottlerocketNode` objects in the cluster.
+    fn all_nodes(&self) -> Vec<BottlerocketNode> {
+        self.brn_reader.state()
+    }
+
+    /// Returns the set of BottlerocketNode objects which is currently being acted upon.
+    ///
+    /// Nodes are being acted upon if they are not in the `WaitingForUpdate` state, or if their desired state does
+    /// not match their current state.
+    #[instrument(skip(self))]
+    fn active_node_set(&self) -> BTreeMap<String, BottlerocketNode> {
+        self.all_nodes()
+            .iter()
+            .filter(|brn| {
+                brn.status.as_ref().map_or(false, |brn_status| {
```

Comment:
Would map_or_else be better here?

Reply:
It's possible I'm misunderstanding your suggestion, but I think `map_or_else` is generally better if the default value is computed by calling a function. If you use `map_or`, the default argument is evaluated when the statement is parsed, which could be wasteful if you end up not needing the value.

Reply:
Here is what I found: 
```Arguments passed to map_or are eagerly evaluated; if you are passing the result of a function call, it is recommended to use map_or_else, which is lazily evaluated.```
I originally thought map_or evaluate both value and map_or_else lazily evaluate the brn_status is Some(), but looks like it applies only on default value.